File: 0701-0800/0761-Special Binary String/0761-Special Binary String.cpp
class Solution {
public:
    string makeLargestSpecial(string S) {
        int n = S.size();
        int count = 0;
        vector<string> strs;
        int i = 0;
        for (int j = 0; j < n; ++j) {
            if (S[j] == '1') {
                ++count;
            }
            else {
                --count;
            }
            
            if (count == 0) {
                strs.push_back('1' + makeLargestSpecial(S.substr(i + 1, j - i - 1)) + '0');
                i = j + 1;
            }
        }
        
        sort(strs.begin(), strs.end(), greater<string>());
        string result;
        for (string str : strs) {
            result += str;
        }
        return result;
    }
};

File: 0701-0800/0761-Special Binary String/0761-Special Binary String.py
class Solution:
    def makeLargestSpecial(self, S: str) -> str:
        strs = []
        start = bal = 0
        for i, c in enumerate(S):
            if c == '1':
                bal += 1
            else:
                bal -= 1
                
            if bal == 0:
                strs.append('1{}0'.format(self.makeLargestSpecial(S[start+1 : i])))
                start = i + 1
        
        return ''.join(sorted(strs, reverse=True))
